File: src/DashSpec.Core/Parsing/IncludeExpander.cs
using DashSpec.Core.Model;

namespace DashSpec.Core.Parsing;

internal static class IncludeExpander
{
    public static void Expand(
        string reference,
        string specDirectory,
        DocumentModuleKind moduleKind,
        ModuleIncludeState state,
        bool tolerateIncompleteIncludes = false)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(reference);
        ArgumentException.ThrowIfNullOrWhiteSpace(specDirectory);

        if (tolerateIncompleteIncludes && IncludeReferenceHeuristics.IsIncomplete(reference))
        {
            return;
        }

        foreach (var path in ResolvePaths(reference, specDirectory))
        {
            ExpandFile(path, specDirectory, moduleKind, state, tolerateIncompleteIncludes);
        }
    }

    private static void ExpandFile(
        string path,
        string specDirectory,
        DocumentModuleKind moduleKind,
        ModuleIncludeState state,
        bool tolerateIncompleteIncludes)
    {
        if (!File.Exists(path))
        {
            throw new FileNotFoundException($"!include not found: '{path}'.", path);
        }

        var extension = Path.GetExtension(path);
        switch (extension.ToLowerInvariant())
        {
            case ".dashlayout":
                AssignLayoutBoard(LayoutModuleParser.ParseLayoutFile(File.ReadAllText(path)), moduleKind, state);
                return;

            case ".dashdiagram":
                RegisterDiagramFile(path, specDirectory, state);
                return;

            case ".dashinclude":
                ExpandDashInclude(path, specDirectory, moduleKind, state, tolerateIncompleteIncludes);
                return;

            case ".dashpresentation":
                RegisterPresentationFile(path, specDirectory, state);
                return;

            case ".dashtransform":
                throw new DashSpecParseException(
                    $"!include '{path}': register transform via .dashdiagram or card block, not module include.");

            default:
                throw new DashSpecParseException(
                    $"!include '{path}': unsupported extension '{extension}'.");
        }
    }

    private static void ExpandDashInclude(
        string path,
        string specDirectory,
        DocumentModuleKind moduleKind,
        ModuleIncludeState state,
        bool tolerateIncompleteIncludes)
    {
        var reader = ParserUtilities.CreateReader(File.ReadAllText(path));
        reader.SkipNewlines();
        if (reader.IsAt(TokenKind.At))
        {
            reader.Advance();
            if (reader.TryKeyword("include"))
            {
                _ = reader.ReadIdent();
                reader.SkipNewlines();
            }
        }

        while (!reader.IsEof)
        {
            if (reader.TryModuleInclude(out var nested))
            {
                Expand(nested, Path.GetDirectoryName(path) ?? specDirectory, moduleKind, state, tolerateIncompleteIncludes);
                reader.SkipNewlines();
                continue;
            }

            if (reader.TryKeyword("layout"))
            {
                var layoutReference = reader.ReadString();
                var layoutPath = SpecIncludeResolver.ResolvePath(layoutReference, specDirectory);
                layoutPath = ResolveLayoutPath(layoutPath);
                AssignLayoutBoard(
                    LayoutModuleParser.ParseLayoutFile(File.ReadAllText(layoutPath)),
                    moduleKind,
                    state);
                reader.SkipNewlines();
                continue;
            }

            if (reader.TryKeyword("diagram"))
            {
                var diagramReference = reader.ReadString();
                var diagramPath = SpecIncludeResolver.ResolvePath(diagramReference, specDirectory);
                diagramPath = ResolveDiagramPath(diagramPath);
                RegisterDiagramFile(diagramPath, specDirectory, state);
                reader.SkipNewlines();
                continue;
            }

            throw reader.Unexpected();
        }
    }

    private static void RegisterDiagramFile(string path, string specDirectory, ModuleIncludeState state)
    {
        var baseDirectory = Path.GetDirectoryName(path) ?? specDirectory;
        var (id, fragment) = DiagramModuleParser.ParseDiagramFileWithId(File.ReadAllText(path), baseDirectory);
        state.RegisterDiagram(id, fragment);
    }

    private static void RegisterPresentationFile(string path, string specDirectory, ModuleIncludeState state)
    {
        var baseDirectory = Path.GetDirectoryName(path) ?? specDirectory;
        var (id, block) = PresentationModuleParser.ParsePresentationFileWithId(
            File.ReadAllText(path),
            baseDirectory);
        state.RegisterChartChromePreset(id, block);
    }

    private static void AssignLayoutBoard(
        LayoutBoardDefinition board,
        DocumentModuleKind moduleKind,
        ModuleIncludeState state)
    {
        if (moduleKind is DocumentModuleKind.Dashboard)
        {
            state.AssignToolbarBoard(board, "!include layout (dashboard toolbar)");
            return;
        }

        state.AssignLayoutBoard(board, "!include layout (tab board)");
    }

    private static IEnumerable<string> ResolvePaths(string reference, string specDirectory)
    {
        if (!reference.Contains('*', StringComparison.Ordinal))
        {
            return [ResolveExistingIncludePath(reference, specDirectory)];
        }

        var combined = Path.GetFullPath(Path.Combine(specDirectory, reference));
        var directory = Path.GetDirectoryName(combined)
            ?? throw new DashSpecParseException($"!include glob has no directory: '{reference}'.");
        var pattern = Path.GetFileName(combined);
        if (string.IsNullOrWhiteSpace(pattern))
        {
            throw new DashSpecParseException($"!include glob requires a file pattern: '{reference}'.");
        }

        if (!Directory.Exists(directory))
        {
            return [];
        }

        return Directory
            .GetFiles(directory, pattern)
            .OrderBy(static path => path, StringComparer.OrdinalIgnoreCase);
    }

    private static string ResolveExistingIncludePath(string reference, string specDirectory)
    {
        var path = SpecIncludeResolver.ResolvePath(reference, specDirectory);
        if (File.Exists(path))
        {
            return path;
        }

        foreach (var ext in new[] { ".dashlayout", ".dashdiagram", ".dashinclude", ".dashpresentation" })
        {
            var withExt = path.EndsWith(ext, StringComparison.OrdinalIgnoreCase) ? path : path + ext;
            if (File.Exists(withExt))
            {
                return withExt;
            }
        }

        return path;
    }

    private static string ResolveLayoutPath(string path) =>
        File.Exists(path) ? path :
        File.Exists(path + ".dashlayout") ? path + ".dashlayout" : path;

    private static string ResolveDiagramPath(string path) =>
        File.Exists(path) ? path :
        File.Exists(path + ".dashdiagram") ? path + ".dashdiagram" : path;
}

File: src/DashSpec.Modeling.Parse/Document/IncludeExpander.fs
namespace DashSpec.Modeling.Parse.Document

open System
open System.IO
open DashSpec.Modeling.Core
open DashSpec.Modeling.Parse
open DashSpec.Modeling.Parse.Diagram
open DashSpec.Modeling.Parse.Include
open DashSpec.Modeling.Parse.Layout
open DashSpec.Modeling.Parse.Lexing
open DashSpec.Modeling.Parse.Presentation
open DashSpec.Modeling.Parse.Tooltip

module IncludeExpander =

    let private isIncomplete (reference: string) =
        if String.IsNullOrWhiteSpace reference then true
        elif reference.Contains '*' then false
        else reference.EndsWith "/" || reference.EndsWith "\\"

    let private resolveExistingIncludePath (reference: string) (specDirectory: string) =
        let path = SpecIncludeResolver.resolvePath reference specDirectory

        if File.Exists path then path
        else
            let extensions =
                [| ".dashlayout"; ".dashdiagram"; ".dashinclude"; ".dashpresentation"; ".dashtooltip" |]

            let mutable resolved = path

            for ext in extensions do
                let withExt =
                    if path.EndsWith(ext, StringComparison.OrdinalIgnoreCase) then path
                    else path + ext

                if File.Exists withExt then resolved <- withExt

            resolved

    let private resolvePaths (reference: string) (specDirectory: string) =
        if not (reference.Contains '*') then
            seq { yield resolveExistingIncludePath reference specDirectory }
        else
            let combined = Path.GetFullPath(Path.Combine(specDirectory, reference))
            let directory = Path.GetDirectoryName combined

            if String.IsNullOrWhiteSpace directory then
                raise (DashSpecParseException($"!include glob has no directory: '{reference}'."))

            let pattern = Path.GetFileName combined

            if String.IsNullOrWhiteSpace pattern then
                raise (DashSpecParseException($"!include glob requires a file pattern: '{reference}'."))

            if not (Directory.Exists directory) then
                Seq.empty
            else
                Directory.GetFiles(directory, pattern)
                |> Array.sortWith (fun left right -> String.Compare(left, right, StringComparison.OrdinalIgnoreCase))
                |> Seq.ofArray

    let private resolveLayoutPath (path: string) =
        if File.Exists path then path
        elif File.Exists(path + ".dashlayout") then path + ".dashlayout"
        else path

    let private resolveDiagramPath (path: string) =
        if File.Exists path then path
        elif File.Exists(path + ".dashdiagram") then path + ".dashdiagram"
        else path

    let private assignLayoutBoard (board: LayoutBoardDefinition) (moduleKind: DocumentModuleKind) (state: ModuleIncludeState) =
        match moduleKind with
        | DocumentModuleKind.Dashboard ->
            state.AssignToolbarBoard(board, "!include layout (dashboard toolbar)")
        | DocumentModuleKind.Tab ->
            state.AssignLayoutBoard(board, "!include layout (tab board)")

    let private registerDiagramFile (path: string) (specDirectory: string) (state: ModuleIncludeState) =
        let baseDirectory =
            Path.GetDirectoryName path |> Option.ofObj |> Option.defaultValue specDirectory

        let id, fragment =
            SpecIncludeFragmentResolver.foldDiagramModuleWithId (File.ReadAllText path) baseDirectory

        state.RegisterDiagram(id, fragment)

    let private registerPresentationFile (path: string) (specDirectory: string) (state: ModuleIncludeState) =
        let baseDirectory =
            Path.GetDirectoryName path |> Option.ofObj |> Option.defaultValue specDirectory

        let text = File.ReadAllText path
        let id, _ = PresentationModuleParser.parsePresentationModuleWithId text
        let block = SpecIncludeFragmentResolver.parsePresentationBlockFromFile text baseDirectory
        state.RegisterChartChromePreset(id, block)

    let private registerTooltipFile (path: string) (state: ModuleIncludeState) =
        let id, definition = TooltipModuleParser.parseTooltipFileWithId (File.ReadAllText path)
        state.RegisterTooltip(id, definition)

    let rec private expandDashInclude
        (path: string)
        (specDirectory: string)
        (moduleKind: DocumentModuleKind)
        (state: ModuleIncludeState)
        (tolerateIncompleteIncludes: bool)
        =
        let reader = ParserUtilities.createReader (File.ReadAllText path)
        reader.SkipNewlines()

        if reader.IsAt TokenKind.At then
            reader.Advance()

            if reader.TryKeyword "include" then
                reader.ReadIdent() |> ignore
                reader.SkipNewlines()

        while not reader.IsEof do
            match reader.TryModuleInclude() with
            | Some nested ->
                expand nested (Path.GetDirectoryName path |> Option.ofObj |> Option.defaultValue specDirectory) moduleKind state tolerateIncompleteIncludes
                reader.SkipNewlines()
            | None ->
                if reader.TryKeyword "layout" then
                    let layoutReference = reader.ReadString()
                    let layoutPath =
                        layoutReference
                        |> fun reference -> SpecIncludeResolver.resolvePath reference specDirectory
                        |> resolveLayoutPath

                    assignLayoutBoard (LayoutModuleParser.parseLayoutFile (File.ReadAllText layoutPath)) moduleKind state
                    reader.SkipNewlines()
                elif reader.TryKeyword "diagram" then
                    let diagramReference = reader.ReadString()

                    let diagramPath =
                        diagramReference
                        |> fun reference -> SpecIncludeResolver.resolvePath reference specDirectory
                        |> resolveDiagramPath

                    registerDiagramFile diagramPath specDirectory state
                    reader.SkipNewlines()
                else
                    raise (reader.Unexpected())

    and private expandFile
        (path: string)
        (specDirectory: string)
        (moduleKind: DocumentModuleKind)
        (state: ModuleIncludeState)
        (tolerateIncompleteIncludes: bool)
        =
        if not (File.Exists path) then
            raise (FileNotFoundException($"!include not found: '{path}'.", path))

        match Path.GetExtension(path).ToLowerInvariant() with
        | ".dashlayout" ->
            assignLayoutBoard (LayoutModuleParser.parseLayoutFile (File.ReadAllText path)) moduleKind state
        | ".dashdiagram" -> registerDiagramFile path specDirectory state
        | ".dashinclude" -> expandDashInclude path specDirectory moduleKind state tolerateIncompleteIncludes
        | ".dashpresentation" -> registerPresentationFile path specDirectory state
        | ".dashtooltip" -> registerTooltipFile path state
        | ".dashtransform" ->
            raise (DashSpecParseException($"!include '{path}': register transform via .dashdiagram or card block, not module include."))
        | extension ->
            raise (DashSpecParseException($"!include '{path}': unsupported extension '{extension}'."))

    and expand
        (reference: string)
        (specDirectory: string)
        (moduleKind: DocumentModuleKind)
        (state: ModuleIncludeState)
        (tolerateIncompleteIncludes: bool)
        =
        if String.IsNullOrWhiteSpace reference then
            invalidArg "reference" "Include reference is required."

        if String.IsNullOrWhiteSpace specDirectory then
            invalidArg "specDirectory" "Spec directory is required."

        if tolerateIncompleteIncludes && isIncomplete reference then
            ()
        else
            for path in resolvePaths reference specDirectory do
                expandFile path specDirectory moduleKind state tolerateIncompleteIncludes
